servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TerminalSignalConsumers.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import static java.util.Objects.requireNonNull;
+
+final class TerminalSignalConsumers implements TerminalSignalConsumer {
+
+    static final class RunnableTerminalSignalConsumer implements TerminalSignalConsumer {
+
+        private final Runnable onFinally;
+
+        RunnableTerminalSignalConsumer(Runnable onFinally) {
+            this.onFinally = requireNonNull(onFinally);
+        }
+
+        @Override
+        public void onComplete() {
+            onFinally.run();
+        }
+
+        @Override
+        public void onError(final Throwable throwable) {
+            onFinally.run();
+        }
+
+        @Override
+        public void onCancel() {
+            onFinally.run();
+        }
+    }
+
+    static final class CompleteTerminalSignalConsumer implements TerminalSignalConsumer {
```

Comment:
`CompleteTerminalSignalConsumer` => `DelegatingTerminalSignalConsumer`?

I am on the fence for wrapping here to make sure that we only invoke the said methods once. Altenratively you can leave the state in respective operators as they are today and if you want to share code of invoking these methods, you can create static methods here like:

```java

static <T> void invokeOnCompleteIfNotDone(TerminalSignalConsumer target, AtomicIntegerFieldUpdater<T> doneUpdater,
                                          final T owner) {
    if (doneUpdater.compareAndSet(owner, 0, 1)) {
        target.onComplete();
    }
}
```

Reply:
I tried this approach first and it works well for `(Before|After)FinallyPublisher`, but similar operators for `Completable` and `Single` use `BeforeCancellable`. It required me to create a new `Cancellable` implementation or pass a lambda to `BeforeCancellable` that will keep references to the state outside of that lambda. Not a big deal but introduces a similar need for wrapping. 
This approach also wraps, but at the same time helps to reduce duplication between operators of maintaining atomic state and allows me to pass only references to the `doFinally::onCancel` for `BeforeCancellable`.

Reply:
Because we have `Single.TerminalSignalConsumer` now this wrapped doesn't make sense anymore. Reverted the previous approach.